src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala

```diff
@@ -681,6 +687,9 @@ class BoxUnbox[BT <: BTypes](val btypes: BT) {
           else if (isPredefAutoUnbox(mi) && typeOK(mi)) BoxKind.checkReceiverPredefLoad(mi, prodCons).map(ModuleGetter(_, mi))
           else None
 
+        case ti: TypeInsnNode if insn.getOpcode == INSTANCEOF =>
+          Some(BoxedPrimitiveTypeCheck(ti, ti.desc == kind.boxClass))
```

Comment:
Yeah, I think there are very few possibilities here (`kind.boxClass`, `Number`, `Comparable`, `Object`, and `Serializable`). But you're right:

```
scala> def f = (1: Integer).isInstanceOf[Object]
f: Boolean

scala> :javap f
/* ... */
  public boolean f();
    descriptor: ()Z
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=1, args_size=1
         0: getstatic     #24                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
         3: iconst_1
         4: invokevirtual #28                 // Method scala/Predef$.int2Integer:(I)Ljava/lang/Integer;
         7: instanceof    #4                  // class java/lang/Object
        10: ireturn
/* ... */
```